java/ql/src/experimental/Security/CWE/CWE-348/UseOfLessTrustedSourceLib.qll

```diff
@@ -0,0 +1,120 @@
+import java
+import DataFlow
+import semmle.code.java.security.QueryInjection
+import experimental.semmle.code.java.Logging
+
+/**
+ * A data flow source of the client ip obtained according to the remote endpoint identifier specified
+ * (`X-Forwarded-For`, `X-Real-IP`, `Proxy-Client-IP`, etc.) in the header.
+ *
+ * For example: `ServletRequest.getHeader("X-Forwarded-For")`.
+ */
+class UseOfLessTrustedSource extends DataFlow::Node {
+  UseOfLessTrustedSource() {
+    exists(MethodAccess ma |
+      ma.getMethod().hasName("getHeader") and
+      ma.getArgument(0).(CompileTimeConstantExpr).getStringValue().toLowerCase() in [
+          "x-forwarded-for", "x-real-ip", "proxy-client-ip", "wl-proxy-client-ip",
+          "http_x_forwarded_for", "http_x_forwarded", "http_x_cluster_client_ip", "http_client_ip",
+          "http_forwarded_for", "http_forwarded", "http_via", "remote_addr"
+        ] and
+      ma = this.asExpr()
+    )
+  }
+}
+
+/** A data flow sink for ip address forgery vulnerabilities. */
+abstract class UseOfLessTrustedSink extends DataFlow::Node { }
+
+/**
+ * A data flow sink for the if condition, which does not include the null judgment of the remote client ip address.
+ *
+ * For example: `if (!StringUtils.startsWith(ipAddr, "192.168.")){...` determine whether the client ip starts
+ * with `192.168.`, and the program can be deceived by forging the ip address.
+ * `if (remoteAddr == null || "".equals(remoteAddr)) {...` judging whether the client ip is a null value,
+ * it needs to be excluded
+ */
+private class IfConditionSink extends UseOfLessTrustedSink {
+  IfConditionSink() {
```

Comment:
There's no need to check this is the condition of an IfStmt any more -- let the sink simply be the argument to `equals` or `contains` or similar, and we can safely assume that e.g. `boolean isEqual = getHeader("X-Forwarded-For", "127.0.0.1")` or whatever will eventually be used in some kind of test.

Reply:
ok